core/src/main/java/ai/timefold/solver/core/impl/domain/variable/cascade/CascadeUpdateVariableListener.java

```diff
@@ -33,12 +35,25 @@ public void beforeVariableChanged(ScoreDirector<Solution_> scoreDirector, Object
     public void afterVariableChanged(ScoreDirector<Solution_> scoreDirector, Object entity) {
         var currentEntity = entity;
         while (currentEntity != null) {
-            scoreDirector.beforeVariableChanged(currentEntity, sourceShadowVariableDescriptor.getVariableName());
-            Object oldValue = sourceShadowVariableDescriptor.getValue(currentEntity);
+            List<Object> oldValueList = new ArrayList<>(sourceShadowVariableDescriptorList.size());
```

Comment:
Personally, I'd introduce two methods for this - one which only has 1 var, and another which handles 2+ vars. The 1-var path will be significantly faster, avoiding the list and the iterations. It is also the most frequent use case, so it makes sense to have a path handling it.